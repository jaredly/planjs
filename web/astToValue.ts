import { IDRef, Loc, RecNode } from 'j3/one-world/shared/nodes';
import { Body, Value } from '../ps/normal/runtime';
import { AST } from './types';
import { asciiToNat } from '../runtime/natToAscii';

type Ctx = {
    processLaw(
        name: string | undefined,
        args: string[],
        lets: { name: string; value: Body }[],
        value: Body,
    ): Value;
    locHash: (loc: Loc) => string;
};

export const findFree = (node: AST, locals: string[], found: string[]) => {
    switch (node.type) {
        case 'app':
            findFree(node.target, locals, found);
            node.args.forEach((n) => findFree(n, locals, found));
            return;
        case 'array':
            node.items.forEach((n) => findFree(n, locals, found));
            return;
        case 'builtin':
        case 'nat':
        case 'pin':
        case 'string':
            return;
        case 'law':
            const fnLocals = node.args.concat(node.lets.map((l) => l.name));
            findFree(node.body, locals.concat(fnLocals), found);
            return;
        case 'local':
            if (!locals.includes(node.name) && !found.includes(node.name)) {
                found.push(node.name);
            }
            return;
    }
};

export const fromRecNode = (node: AST, locals: string[], ctx: Ctx): Body => {
    switch (node.type) {
        case 'app': {
            let res = fromRecNode(node.target, locals, ctx);
            for (let i = 0; i < node.args.length; i++) {
                res = [0, res, fromRecNode(node.args[i], locals, ctx)];
            }
            return res;
        }
        case 'array':
            throw new Error('not yet');
        case 'nat':
            return node.number;
        case 'builtin':
            return node.name;
        case 'law':
            const fnLocals = [
                node.name?.text ?? 'self',
                ...node.args,
                ...node.lets.map((l) => l.name),
            ];
            let extraArgs: string[] = [];
            findFree(node.body, fnLocals, extraArgs);
            // assume everything else is global
            extraArgs = extraArgs.filter((f) => locals.includes(f));
            const args = [
                node.name?.text ?? 'self',
                ...extraArgs,
                ...node.args,
            ];
            const allLocals = args.concat(node.lets.map((l) => l.name));
            // console.log('law', args, allLocals, extraArgs);
            let res = ctx.processLaw(
                node.name?.text,
                args,
                node.lets.map((l) => ({
                    name: l.name,
                    value: fromRecNode(l.value, allLocals, ctx),
                })),
                fromRecNode(node.body, allLocals, ctx),
            );
            for (let arg of extraArgs) {
                if (locals.indexOf(arg) === -1)
                    throw new Error(`unbound free vbl ${arg}`);
                res = [0, res, locals.indexOf(arg)];
            }
            return res;
        case 'local':
            const at = locals.indexOf(node.name);
            if (at === -1) {
                // console.log('n', locals);
                // throw new Error(`unbound local ${node.name}`);
                return node.name;
            }
            return [3, at];
        case 'string':
            if (node.templates.length) throw new Error('not supported tpl yet');
            return asciiToNat(node.first);
        case 'pin':
            return ctx.locHash(node.ref);
    }
};
