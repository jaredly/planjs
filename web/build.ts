import { watch } from 'fs';
import { join } from 'path';

const bounce = (time: number, fn: () => unknown) => {
    let wait: null | Timer = null;
    return () => {
        if (wait != null) clearTimeout(wait);
        wait = setTimeout(() => fn(), time);
    };
};

let edited: string[] = [];
const rebuild = bounce(10, () => {
    console.log('rebuilding for', edited);
    edited = [];
    Promise.all([
        Bun.build({
            entrypoints: ['./web/run.tsx'],
            outdir: './web',
            naming: 'run.js',
        }),
        // Bun.build({
        //     entrypoints: ['./one-world/client/cli/xterm/run.ts'],
        //     outdir: './',
        //     naming: 'run.js',
        // }),
    ])
        .then(([one]) => {
            console.log(one.logs);
        })
        .catch((err) => {
            console.log('failed? idk');
        });
});

const ignore = ['.git/', 'node_modules/', 'worker.js', 'run.js'];

watch('.', { recursive: true }, (event, filename) => {
    if (!filename) return;
    if (ignore.some((n) => filename.startsWith(n))) {
        // ignore
        return;
    }
    if (filename.match(/\.tsx?$/)) {
        edited.push(filename);
        rebuild();
    } else {
        console.log('ignore', filename);
    }
});

const service = Bun.serve({
    async fetch(req) {
        let pathname = new URL(req.url).pathname;
        if (pathname.endsWith('/')) {
            pathname += 'index.html';
        }
        const file = Bun.file(join('./web', pathname));
        return new Response(file);
    },
});

console.log(`Serving http://${service.hostname}:${service.port}`);
