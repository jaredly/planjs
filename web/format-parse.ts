import { CTX } from 'j3/one-world/evaluators/simplest';
import { Id, IDRef, Loc, RecNode } from 'j3/one-world/shared/nodes';
import { AST, Let } from './types';
import { asciiToNat } from '../runtime/natToAscii';

export const getExport = (top: RecNode): Id<Loc> | void => {
    if (
        top.type === 'list' &&
        top.items.length >= 2 &&
        top.items[0].type === 'id' &&
        (top.items[0].text === 'defn' || top.items[0].text === 'def') &&
        top.items[1].type === 'id'
    ) {
        return top.items[1];
    }
};

const parseLet = (ctx: CTX, node: RecNode): Let | null => {
    if (
        node.type === 'list' &&
        node.items.length === 3 &&
        node.items[0].type === 'id' &&
        node.items[0].text === 'let' &&
        node.items[1].type === 'id'
    ) {
        const value = parseExpr(ctx, node.items[2]);
        return value ? { name: node.items[1].text, value } : null;
    }
    ctx.errors.push({ loc: node.loc, text: 'bad let' });
    return null;
};

const topForms: Record<
    string,
    (ctx: CTX, loc: Loc, ...args: RecNode[]) => AST | void
> = {
    // TODO: defs can have lets
    def(ctx, loc, name, ...body) {
        if (!name || !body.length || name.type !== 'id' || name.ref) {
            ctx.errors.push({ loc, text: 'bad def form' });
            return;
        }
        const lets: Let[] = [];
        for (let i = 0; i < body.length - 1; i++) {
            const l = parseLet(ctx, body[i]);
            if (l) lets.push(l);
        }
        const expr = parseExpr(ctx, body[body.length - 1]);
        ctx.exports?.push({ kind: 'value', loc: name.loc });
        return expr
            ? {
                  type: 'law',
                  name,
                  args: [],
                  lets,
                  loc,
                  body: expr,
              }
            : undefined;
    },
    defn(ctx, loc, name, args, ...body) {
        if (!name || !args || !body.length || name.type !== 'id') {
            console.log(name, args, body);
            ctx.errors.push({ loc, text: 'bad defn form' });
            return;
        }
        ctx.layouts[getLoc(loc)] = {
            type: 'vert',
            layout: { tightFirst: 3, indent: 4 },
        };

        ctx.exports?.push({ kind: 'value', loc: name.loc });
        const fn = forms.fn(ctx, loc, args, ...body);
        return fn?.type === 'law' ? { ...fn, name } : undefined;
    },
    // defmacro(ctx, loc, name, args, value) {
    //     if (
    //         !name ||
    //         !args ||
    //         !value ||
    //         name.type !== 'id' ||
    //         name.ref ||
    //         args.type !== 'array' ||
    //         !args.items.every((item) => item.type === 'id')
    //     ) {
    //         ctx.errors.push({ loc, text: 'bad form' });
    //         return;
    //     }
    //     ctx.exports?.push({ kind: 'macro', loc: name.loc });
    //     const body = parseExpr(ctx, value);
    //     return body
    //         ? {
    //               type: 'defmacro',
    //               args: args.items.map((arg) => (arg as Id<Loc>).text),
    //               loc: name.loc,
    //               body,
    //           }
    //         : undefined;
    // },
};

const getLoc = (l: Loc) => l[l.length - 1][1];

const forms: Record<
    string,
    (ctx: CTX, loc: Loc, ...args: RecNode[]) => AST | void
> = {
    fn(ctx, loc, args, ...body) {
        if (
            !args ||
            !body.length ||
            args.type !== 'array' ||
            !args.items.every((arg) => arg.type === 'id')
        ) {
            ctx.errors.push({ loc, text: 'bad fn form' });
            return;
        }
        const lets: Let[] = [];
        for (let i = 0; i < body.length - 1; i++) {
            const l = parseLet(ctx, body[i]);
            if (l) lets.push(l);
        }
        const expr = parseExpr(ctx, body[body.length - 1]);
        return expr
            ? {
                  type: 'law',
                  lets,
                  args: args.items.map((arg) => (arg as Id<Loc>).text),
                  body: expr,
                  loc,
              }
            : undefined;
    },
};

export const fixGlobals = (node: RecNode, globals: Record<string, IDRef>) => {
    switch (node.type) {
        case 'list':
        case 'record':
        case 'array':
            node.items.forEach((n) => fixGlobals(n, globals));
            return;
        case 'string':
            node.templates.forEach((t) => fixGlobals(t.expr, globals));
            return;
        case 'table':
            node.rows.forEach((r) => r.forEach((c) => fixGlobals(c, globals)));
            return;
        case 'id':
            if (!node.ref && globals[node.text]) {
                const ref = globals[node.text];
                if (ref.type === 'toplevel' && ref.loc === node.loc) {
                    return;
                }
                node.ref = globals[node.text];
            }
            return;
    }
};

const parseExpr = (ctx: CTX, value: RecNode): AST | void => {
    switch (value.type) {
        case 'string':
            if (value.templates.length) {
                return;
            }
            const templates = [];
            for (let item of value.templates) {
                const expr = parseExpr(ctx, item.expr);
                if (!expr) return;
                templates.push({ expr, suffix: item.suffix });
            }
            return {
                type: 'string',
                first: value.first,
                templates,
                loc: value.loc,
            };
        case 'array': {
            const items: AST[] = [];
            for (let item of value.items) {
                const parsed = parseExpr(ctx, item);
                if (!parsed) return;
                items.push(parsed);
            }
            return { type: 'array', items, loc: value.loc };
        }
        case 'id':
            if (value.ref?.type === 'toplevel') {
                ctx.references.push({ loc: value.loc, ref: value.ref });
                return {
                    type: 'pin',
                    ref: value.ref.loc,
                    loc: value.loc,
                };
            }
            if (value.ref?.type === 'builtin') {
                ctx.references.push({ loc: value.loc, ref: value.ref });
                return {
                    type: 'builtin',
                    name: value.text,
                    // kind: value.ref.kind,
                    loc: value.loc,
                };
            }
            if (value.text.length > 0) {
                const num = Number(value.text);
                if (Number.isInteger(num)) {
                    return { type: 'nat', number: BigInt(num), loc: value.loc };
                }
                // if (Number.isFinite(num)) {
                //     return { type: 'float', value: num };
                // }
            }
            if (ctx.cursor != null && ctx.cursor === value.loc[0][1]) {
                ctx.autocomplete = {
                    local: [],
                    kinds: ['kwd', 'value'],
                };
                // } else if (ctx.cursor != null) {
                //     console.log('an id', ctx.cursor, value.loc[0][1]);
            }
            if (value.text === '') {
                ctx.errors.push({ loc: value.loc, text: 'blank' });
                return;
            }
            return { type: 'local', loc: value.loc, name: value.text };

        case 'list':
            if (
                value.type === 'list' &&
                value.items.length > 0 &&
                value.items[0].type === 'id'
            ) {
                const id = value.items[0].text;
                if (forms[id]) {
                    return forms[id](ctx, value.loc, ...value.items.slice(1));
                }
            }
            if (value.items.length === 0) {
                return { type: 'builtin', name: 'nil', loc: value.loc };
            }
            if (
                value.items.length === 1 &&
                value.items[0].type === 'id' &&
                !value.items[0].ref &&
                value.items[0].text === ''
            ) {
                return { type: 'builtin', name: 'nil', loc: value.loc };
            }
            if (value.items.length > 1) {
                const target = parseExpr(ctx, value.items[0]);
                const args = value.items
                    .slice(1)
                    .filter(
                        (t) => !(t.type === 'id' && !t.ref && t.text === ''),
                    )
                    .map((arg) => parseExpr(ctx, arg));
                return target && args.every((arg) => !!arg)
                    ? {
                          type: 'app',
                          target,
                          args: args as AST[],
                          loc: value.loc,
                      }
                    : undefined;
            }
            return parseExpr(ctx, value.items[0]);
    }
    throw new Error(`invalid expr`);
};

export const parseTop = (ctx: CTX, node: RecNode): AST | null => {
    if (node.type === 'rich-block') {
        return null;
    }
    try {
        if (
            node.type === 'list' &&
            node.items.length > 0 &&
            node.items[0].type === 'id'
        ) {
            const id = node.items[0].text;
            if (topForms[id]) {
                if (
                    ctx.cursor != null &&
                    node.items[0].loc[0][1] === ctx.cursor &&
                    node.items.length === 1
                ) {
                    ctx.autocomplete = { kinds: ['kwd'], local: [] };
                }
                return (
                    topForms[id](ctx, node.loc, ...node.items.slice(1)) ?? null
                );
            }
        }
        const expr = parseExpr(ctx, node);
        return expr ?? null;
    } catch (err) {
        ctx.errors.push({ loc: node.loc, text: (err as Error).message });
        return null;
    }
};

export const parse = (node: RecNode) => {
    const ctx: CTX = {
        layouts: {},
        styles: {},
        exports: [],
        errors: [],
        tableHeaders: {},
        autocomplete: undefined,
        references: [],
        // cursor:,
    };
    const top = parseTop(ctx, node);
    return { ...ctx, top };
};
