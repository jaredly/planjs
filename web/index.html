<html>
    <head>
        <link rel="favorite icon" href="./favicon.png" />
    </head>
    <body>
        <script></script>
        <style>
            body {
                background-color: #111 !important;
                color: white;
            }
            .xterm-viewport {
                background-color: #111 !important;
            }
        </style>
        <script src="./run.js"></script>
    </body>
</html>
