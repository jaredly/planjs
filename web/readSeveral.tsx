import { readerMulti } from 'j3/one-world/evaluators/boot-ex/reader';
import { RecNode, IDRef, keyForLoc } from 'j3/one-world/shared/nodes';
import { getExport, parse } from './format-parse';
import { AST } from './types';

export const readSeveral = (fullText: string) => {
    const tops: RecNode[] = [];
    let i = 0;

    const globals: Record<string, IDRef> = {
        NCASE: { type: 'builtin', kind: '' },
        INC: { type: 'builtin', kind: '' },
        PCASE: { type: 'builtin', kind: '' },
        LAW: { type: 'builtin', kind: '' },
        PIN: { type: 'builtin', kind: '' },
    };

    while (i < fullText.length) {
        const result = readerMulti(i, fullText, 'full', globals, false);
        if (!result) break;
        tops.push(result.node);
        i = result.i;
    }
    const nameForLoc: Record<string, string> = {};

    tops.forEach((top) => {
        const exp = getExport(top);
        if (exp) {
            globals[exp.text] = {
                type: 'toplevel',
                kind: '',
                loc: exp.loc,
            };
            nameForLoc[keyForLoc(exp.loc)] = exp.text;
        }
    });

    // tops.forEach((top) => fixGlobals(top, globals));

    const parseds: AST[] = [];

    tops.forEach((top, i) => {
        const parsed = parse(top, globals);
        if (!parsed.top) {
            throw new Error(
                `cant parse I guess ${JSON.stringify(parsed.errors)}`,
            );
        }
        parseds.push(parsed.top);
    });

    return { tops, parseds, nameForLoc, globals };
};
