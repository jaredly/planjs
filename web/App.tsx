import React, { useEffect, useMemo, useState } from 'react';
import { compileMain, runMain, Toplevels } from '../ps/normal/compile';
import { Controlled as CodeMirror } from 'react-codemirror2';
// import { clojure } from '@nextjournal/lang-clojure';
require('codemirror/mode/clojure/clojure');

// const builtins = `
// (defn + [a b]
//     (NCASE b (fn [a] (INC (+ a b))) a))
// `;

const initialText = `
(defn ! [v _] v)
(defn + [a b]
    (NCASE b (fn [a] (INC (+ a b))) a))
(defn lcase [lst nil cons]
    (PCASE
        (! nil)
        (! nil)
        cons
        (! nil)
        lst))
(defn zip [f one two]
    (lcase one
        0
        (fn [a one]
            (lcase two
                0
                (fn [b two]
                    ((f a b) (zip f one two)))))))
(defn drop [n lst]
    (NCASE
        lst
        (fn [n_]
            (lcase lst lst
                (fn [a rest] (drop n_ rest))))
        n))
(defn take [n lst]
    (NCASE 0 (fn [n_]
        (lcase lst
            0
            (fn [head tail] (head (take n_ tail)))))
        n))
(defn fib [n]
    (let [self   (0 (1 (zip + self offset)))
          offset (drop 1 self)]
        (take n self)))
(def main (fib 10))
`;

const key = 'plan:sandbox';

import { ABlock } from 'j3/one-world/shared/ir/block-to-attributed-text';
import { rgb } from './showMore';

export const aBlockToNodes = (block: ABlock) => {
    return block.map((line, l) => (
        <div key={l}>
            {line.map((chunk, i) => (
                <span
                    key={`${l}:${i}`}
                    style={{
                        ...chunk.style,
                        color: chunk.style?.color
                            ? rgb(chunk.style.color)
                            : undefined,
                        background: chunk.style?.background
                            ? rgb(chunk.style.background)
                            : undefined,
                        textDecorationStyle: 'dotted',
                    }}
                >
                    {chunk.text}
                </span>
            ))}
        </div>
    ));
};

import { recNodeToText } from 'j3/one-world/client/cli/drawDocNode';
import { keyForLoc } from 'j3/one-world/shared/nodes';
import { forceDeep, show } from '../ps/normal/runtime';
import { setupStepper } from '../ps/stepper/setupStepper';
import { showMValue } from '../ps/stepper/showMValue';
import { fromAST } from './astToValue';
import { parse } from './format-parse';
import { readSeveral } from './readSeveral';

export const App = () => {
    const [text, setText] = useState<string>(localStorage[key] || initialText);
    const [v, setV] = useState(50);

    useEffect(() => {
        localStorage[key] = text;
    }, [text]);

    const compiled = useMemo(() => {
        const maxWidth = 50;
        try {
            let fullText = text;
            const { tops, parseds, nameForLoc, globals } =
                readSeveral(fullText);

            const topValues: Toplevels = {};

            const full: JSX.Element[] = [];
            tops.forEach((top, i) => {
                const parsed = parse(top, globals);
                if (!parsed.top) {
                    throw new Error(
                        `cant parse I guess ${JSON.stringify(parsed.errors)}`,
                    );
                }
                // parseds.push(parsed.top);
                let base =
                    (parsed.top.type === 'law' && parsed.top.name?.text) ||
                    `top${i}`;
                // base = clean(base);
                const body = fromAST(parsed.top, [], {
                    locHash(loc) {
                        return nameForLoc[keyForLoc(loc)];
                        // return loc.map((l) => `${l[0]}_${l[1]}`).join('__');
                    },
                    lawNum: 0,
                    processLaw(fns, name, args, lets, value) {
                        let hash =
                            fns === 0 ? base : `${base}_${fns}_${name ?? ''}`;

                        topValues[hash] = {
                            type: 'law',
                            name: hash,
                            args,
                            lets,
                            body: value,
                        };

                        return hash;
                    },
                });
                if (parsed.top.type !== 'law') {
                    topValues[base] = { type: 'plain', value: body };
                }

                const text = recNodeToText(
                    top,
                    parsed,
                    maxWidth,
                    (loc) => nameForLoc[keyForLoc(loc)],
                );
                full.push(
                    <div style={{ marginBottom: 16 }} key={i}>
                        {aBlockToNodes(text)}
                    </div>,
                );
            });

            const stepper = setupStepper(parseds);

            let res = '';
            let slider = false;

            const code = compileMain(topValues);
            // res = code; // show(result);
            try {
                const result = runMain(code);
                if (typeof result === 'function') {
                    if (result.length === 0) {
                        res += '\n\n' + show(forceDeep(result()));
                    } else {
                        res += '\n\n' + show(forceDeep(result(BigInt(v))));
                        slider = true;
                    }
                }
            } catch (e) {
                const err = e as Error;

                res += `\n\n${err.message}\n${err.stack}`;
            }

            return { full, res, slider, code, stepper };
        } catch (err) {
            return {
                full: (err as Error).message + '\n' + (err as Error).stack,
                res: '',
                slider: false,
            };
        }
    }, [text, v]);

    return (
        <div
            style={{
                width: '100vw',
                height: '100vh',
                display: 'flex',
                flexDirection: 'row',
            }}
        >
            <div
                style={{
                    // flex: 1,
                    alignSelf: 'stretch',
                    display: 'flex',
                    flexDirection: 'column',
                }}
            >
                <CodeMirror
                    value={text}
                    // height="200px"
                    // extensions={[clojure()]}
                    options={{
                        mode: 'clojure',
                        theme: 'material-ocean',
                        lineNumbers: true,
                    }}
                    onBeforeChange={(_, __, value) => {
                        console.log(value);
                        setText(value);
                        //
                    }}
                    onChange={(editor, data, v) => {
                        console.log(v);
                        // setText(v);
                    }}
                />
            </div>
            {/* <textarea
                style={{
                    flex: 1,
                    // width: 500,
                    // height: 800,
                }}
                value={text}
                onChange={(e) => setText(e.target.value)}
            /> */}
            <div
                style={{
                    flex: 1,
                    whiteSpace: 'pre-wrap',
                    padding: 16,
                    fontFamily: 'monospace',
                    minWidth: 0,
                }}
            >
                <div style={{ display: 'flex' }}>
                    <div style={{ flex: 1 }}>{compiled.full}</div>
                    {/* <div
                        style={{
                            flex: 1,
                            minHeight: 0,
                            overflow: 'auto',
                            // height: 500,
                        }}
                    >
                        {compiled.code}
                    </div> */}
                </div>
                <br />
                {compiled.stepper ? (
                    <>
                        <table>
                            <tbody>
                                {compiled.stepper.memory.heap.map((v, i) => (
                                    <tr key={i}>
                                        <td>{i}</td>
                                        <td>{showMValue(v)}</td>
                                    </tr>
                                ))}
                            </tbody>
                        </table>
                        {Object.entries(compiled.stepper.memory.laws).map(
                            ([name, { buffer, arity }]) => (
                                <div key={name} style={{ padding: 16 }}>
                                    <div>
                                        {name} (arity {arity})
                                    </div>
                                    <table>
                                        <tbody>
                                            {buffer.map((v, i) => (
                                                <tr key={i}>
                                                    <td>{i}</td>
                                                    <td>{showMValue(v)}</td>
                                                </tr>
                                            ))}
                                        </tbody>
                                    </table>
                                </div>
                            ),
                        )}
                        {/* {JSON.stringify(compiled.stepper.laws, (_, v) =>
                            typeof v === 'bigint' ? v + '' : v,
                        )} */}
                    </>
                ) : null}
                <br />
                {compiled.slider ? (
                    <>
                        <input
                            type="range"
                            min="0"
                            max="200"
                            style={{ width: 600 }}
                            value={v}
                            onChange={(e) => setV(+e.target.value)}
                        />
                        {v}
                    </>
                ) : null}
                <br />
                {compiled.res}
            </div>
        </div>
    );
};
