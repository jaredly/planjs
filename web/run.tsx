import React from 'react';
import { createRoot, Root } from 'react-dom/client';
import { App } from './App';

declare global {
    interface Window {
        reactRoot?: Root;
    }
}

const rootNode = () => {
    if (window.reactRoot) return window.reactRoot;
    const node = document.createElement('div');
    document.body.append(node);
    return (window.reactRoot = createRoot(node));
};

rootNode().render(<App />);
console.log('hi');
