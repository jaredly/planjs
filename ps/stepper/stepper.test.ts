import { test, expect } from 'bun:test';
import { readSeveral } from '../../web/readSeveral';
import {
    deep,
    Memory,
    MValue,
    prepareLaw,
    prettyMValue,
    showHeap,
    showMValue,
    stackMain,
    step,
} from './runtime';

const setup = (input: string) => {
    const { parseds } = readSeveral(input);
    const { memory } = stackMain(parseds);
    if (!memory.laws.main) {
        throw new Error(`no main law`);
    }
    const main = prepareLaw(memory.laws.main.buffer, [], memory.heap.length);
    memory.heap.push(...main);
    const dest = memory.heap.length - 1;
    memory.stack.push({ at: dest });

    return { dest, memory };
};

const check = (input: string, output: string, monitor = false) => {
    const { dest, memory } = setup(input);

    if (monitor) {
        expect(
            Object.entries(memory.laws)
                .map(
                    ([name, { buffer, arity }]) =>
                        `LAW ${name} (arity ${arity})\n${showHeap(buffer)}`,
                )
                .join('\n\n'),
        ).toMatchSnapshot();
        expect('Initial Heap:\n' + showHeap(memory.heap)).toMatchSnapshot();
    }

    const logs: string[] = [];
    const log = () => {
        if (monitor) {
            logs.push(debugMem(memory));
        }
    };
    const slog = monitor
        ? (...args: any[]) =>
              logs.push(
                  args
                      .map((m) =>
                          typeof m === 'string' ? m : JSON.stringify(m),
                      )
                      .join(' '),
              )
        : undefined;

    log();
    for (let i = 0; i < 20 && memory.stack.length; i++) {
        try {
            step(memory, slog);
        } catch (err) {
            expect(logs.join('\n\n')).toMatchSnapshot();
            throw err;
        }
        if (monitor) {
            logs.push(prettyMValue(memory.heap[dest], memory));
        }
        log();
    }

    if (monitor) {
        logs.push(`-- finished stack -------------------------------`);
        logs.push(prettyMValue(memory.heap[dest], memory));
    }

    if (memory.stack.length) {
        if (monitor) {
            expect(logs.join('\n\n')).toMatchSnapshot();
        }
        // console.log(logs);
        throw new Error('why is there morestack - before deep');
    }

    let depth = 0;
    while (deep(memory, dest) && depth++ < 20) {
        logs.push(JSON.stringify(memory.stack));
        for (let i = 0; i < 20 && memory.stack.length; i++) {
            step(memory);
            log();
        }
    }

    if (monitor) {
        expect(logs.join('\n\n')).toMatchSnapshot();
    }

    if (memory.stack.length) {
        // console.log(logs);
        throw new Error('why is there morestack - after deep');
    }

    expect(prettyMValue(memory.heap[dest], memory)).toEqual(output);
};

test('just a number', () => {
    check(`(def main 30)`, '30');
});

test('inc it up', () => {
    check(`(def main (INC 29))`, '30');
});

test('call a fn', () => {
    check(
        `
        (defn hi [x] (INC (INC x)))
        (def main (hi 29))`,
        '31',
    );
});

test('call a fn with two args', () => {
    check(
        `
        (defn hi [x y] (INC (INC x)))
        (def main (hi 29 1))`,
        '31',
    );
});

test('call a fn with two args', () => {
    check(
        `
        (defn hi [x y] (INC (INC y)))
        (def main (hi 29 1))`,
        '3',
    );
});

test('lets do an apppp', () => {
    check(`(def main (1 2 3))`, '(1 2 3)');
});

test('do we force deep', () => {
    check(`(def main (1 (INC 1) (INC (INC 1))))`, '(1 2 3)');
});

test('a little ncase as a treat', () => {
    check(`(def main (NCASE 2 10 0))`, '2');
    // check(`(def main (NCASE 2 INC 5))`, '5');
    // check(`(def main (NCASE 2 10 (1 1)))`, '2');
});

const debugMem = (mem: Memory) => {
    return (
        showHeap(mem.heap) +
        '\n---\n' +
        mem.stack
            .map(
                (frame, i) =>
                    `${i.toString().padStart(2, ' ')}: ${frame.at}${
                        frame.step ? ' ' + frame.step : ''
                    }`,
            )
            .join('\n')
    );
};

test('ncase with nonzero tho', () => {
    check(`(def main (NCASE 2 INC 5))`, '5');
});

test('ncase with madness', () => {
    check(`(def main (NCASE 2 10 (1 1)))`, '2');
});

test(`nested fn`, () => {
    check(`(def main ((fn [x] (x x)) 2))`, '(2 2)');
});

test(`nested fn w/ scope`, () => {
    check(
        `
        (defn doot [z] ((fn [x] (z x)) 23))
        (def main (doot 2))`,
        '(2 23)',
    );
});

test('can we please plus', () => {
    check(
        `
    (defn + [a b] (NCASE b (fn [a_] (INC (+ a_ b))) a))
    (def main (+ 2 3))
        `,
        '5',
    );
});

test('now for pcase(n) if you pclease', () => {
    check(`(def main (PCASE 0 1 2 3 0))`, '(3 0)');
});

test('now for pcase(a) if you pclease', () => {
    check(`(def main (PCASE 7 8 9 10 (5 6)))`, '(9 5 6)');
});

test('lcase', () => {
    check(
        `
(defn ! [v _] v)
(defn lcase [lst nil cons]
    (PCASE (! nil) (! nil) cons (! nil) lst))

(def main (lcase (1 2) 0 5))
        `,
        '(5 1 2)',
    );

    check(
        `
(defn ! [v _] v)
(defn lcase [lst nil cons]
    (PCASE (! nil) (! nil) cons (! nil) lst))

(def main (lcase 2 7 5))
        `,
        '7',
    );
});

test('take', () => {
    check(
        `
(defn ! [v _] v)
(defn lcase [lst nil cons]
    (PCASE (! nil) (! nil) cons (! nil) lst))
(defn take [n lst]
    (NCASE 0 (fn [n_]
        (lcase lst
            0
            (fn [head tail] (head (take n_ tail)))))
        n))

(def main (take 3 (1 (2 (3 (4 5))))))
        `,
        '(1 (2 (3 0)))',
    );
});

test('tinf', () => {
    check(
        `
(defn ! [v _] v)
(defn lcase [lst nil cons]
    (PCASE (! nil) (! nil) cons (! nil) lst))
(defn take [n lst]
    (NCASE 0 (fn [n_]
        (lcase lst
            0
            (fn [head tail] (head (take n_ tail)))))
        n))

(def main
    (let inf (5 inf))
    (take 3 inf))
        `,
        '(5 (5 (5 0)))',
        true,
    );
});

test.skip('annnnnd now like ... maybe fibonacci', () => {
    check(
        `
(defn + [a b]
    (NCASE b (fn [a] (INC (+ a b))) a))

(defn ! [v _] v)

(defn lcase [lst nil cons]
    (PCASE
        (! nil)
        (! nil)
        cons
        (! nil)
        lst))

(defn zip [f one two]
    (lcase one
        0
        (fn [a one]
            (lcase two
                0
                (fn [b two]
                    ((f a b) (zip f one two)))))))
(defn drop [n lst]
    (NCASE
        lst
        (fn [n_]
            (lcase lst lst
                (fn [a rest] (drop n_ rest))))
        n))
(defn take [n lst]
    (NCASE 0 (fn [n_]
        (lcase lst
            0
            (fn [head tail] (head (take n_ tail)))))
        n))

(defn fib [n]
    (let self (0 (1 (zip + self offset))))
    (let offset (drop 1 self))
    (take n self))

(defn inf [x] (let self (x self)) self)

(defn main [n] (take 5 (inf 2)))

        `,
        `(2 2 2 2 2)`,
    );
});
