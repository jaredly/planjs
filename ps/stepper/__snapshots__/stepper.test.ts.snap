// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`call a fn 1`] = `""`;

exports[`call a fn with two args 1`] = `""`;

exports[`call a fn with two args 2`] = `""`;

exports[`nested fn w/ scope 1`] = `""`;

exports[`can we please plus 1`] = `""`;

exports[`lcase 1`] = `""`;

exports[`take 1`] = `""`;

exports[`tinf 1`] = `""`;

exports[`annnnnd now like ... maybe fibonacci 1`] = `""`;
